[BUG] stanc.js persisting model name & namespace between calls

### Current Behavior:

After calling `stanc.js` once with a model name, all future calls will reuse the same initial model name:

```r
stanc_js <- tempfile(fileext = ".js")
download.file("https://github.com/stan-dev/stanc3/releases/download/nightly/stanc.js",
              destfile = stanc_js)

qjs_ctx <- QuickJSR::JSContext$new()
qjs_ctx$source(stanc_js)

mod1 <- "data { real y_mean; } parameters { real y; } model { y ~ normal(y_mean, 1); }"
```

```r
> qjs_res_1 <- qjs_ctx$call("stanc", "model_one", mod1)
> cat(qjs_res_1$result, sep = "\n")
# // Code generated by stanc v2.36.0-28-gd13e71e
# #include <stan/model/model_header.hpp>
# namespace model_one_model_namespace {
# ...
```

```r
> qjs_res_2 <- qjs_ctx$call("stanc", "model_two", mod1)
> cat(qjs_res_2$result, sep = "\n")
# // Code generated by stanc v2.36.0-28-gd13e71e
# #include <stan/model/model_header.hpp>
# namespace model_one_model_namespace {
# ...
```

(happens with both V8 & QuickJS)

This is an issue for R packages since it leads to duplicate symbols, so I'm seeing failures like this for `rstanarm`:
```
C:\rtools44\x86_64-w64-mingw32.static.posix\bin/ld.exe: stanExports_binomial.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_binomial.h:29: multiple definition of `model_bernoulli_model_namespace::profiles__[abi:cxx11]'; stanExports_bernoulli.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_bernoulli.h:29: first defined here
C:\rtools44\x86_64-w64-mingw32.static.posix\bin/ld.exe: stanExports_continuous.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_continuous.h:29: multiple definition of `model_bernoulli_model_namespace::profiles__[abi:cxx11]'; stanExports_bernoulli.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_bernoulli.h:29: first defined here
C:\rtools44\x86_64-w64-mingw32.static.posix\bin/ld.exe: stanExports_count.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_count.h:29: multiple definition of `model_bernoulli_model_namespace::profiles__[abi:cxx11]'; stanExports_bernoulli.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_bernoulli.h:29: first defined here
C:\rtools44\x86_64-w64-mingw32.static.posix\bin/ld.exe: stanExports_jm.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_jm.h:29: multiple definition of `model_bernoulli_model_namespace::profiles__[abi:cxx11]'; stanExports_bernoulli.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_bernoulli.h:29: first defined here
C:\rtools44\x86_64-w64-mingw32.static.posix\bin/ld.exe: stanExports_lm.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_lm.h:29: multiple definition of `model_bernoulli_model_namespace::profiles__[abi:cxx11]'; stanExports_bernoulli.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_bernoulli.h:29: first defined here
C:\rtools44\x86_64-w64-mingw32.static.posix\bin/ld.exe: stanExports_mvmer.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_mvmer.h:29: multiple definition of `model_bernoulli_model_namespace::profiles__[abi:cxx11]'; stanExports_bernoulli.o:C:\Users\Andrew\Git\rstanarm\src/stanExports_bernoulli.h:29: first defined here
```

--- a/src/driver/Entry.ml
+++ b/src/driver/Entry.ml
@@ -12,22 +12,27 @@ let fmt_sexp s =
   Sexp.pp_hum ppf s;
   Format.flush_str_formatter ()
 
-let mangle =
-  String.concat_map ~f:(fun c ->
-      Char.(
-        if is_alphanum c || c = '_' then to_string c
-        else match c with '-' -> "_" | _ -> "x" ^ Int.to_string (to_int c)))
+let set_model_name model_name =
+  let mangle =
+    String.concat_map ~f:(fun c ->
+        Char.(
+          if is_alphanum c || c = '_' then to_string c
+          else match c with '-' -> "_" | _ -> "x" ^ Int.to_string (to_int c)))
+  in
+  let model_name_munged =
+    Flags.remove_dotstan List.(hd_exn (rev (String.split model_name ~on:'/')))
+  in
+  if String.equal model_name model_name_munged then
+    (* model name was not file-like, so we leave as is (e.g. from --name argument) *)
+    Typechecker.model_name := mangle model_name
+  else
+    (* model name was a file-like thing, so we add _model to match existing behavior *)
+    Typechecker.model_name := mangle (model_name_munged ^ "_model")
 
 let reset_mutable_states model_name (flags : Flags.t) =
   Common.Gensym.reset_danger_use_cautiously ();
   Include_files.include_provider := flags.include_source;
-  if String.equal !Typechecker.model_name "" then
-    Typechecker.model_name :=
-      mangle
-        (Flags.remove_dotstan
-           List.(hd_exn (rev (String.split model_name ~on:'/'))))
-      ^ "_model"
-  else Typechecker.model_name := mangle !Typechecker.model_name;
+  set_model_name model_name;
   Typechecker.check_that_all_functions_have_definition :=
     not flags.allow_undefined;
   Transform_Mir.use_opencl := flags.use_opencl;

--- a/src/stanc/stanc.ml
+++ b/src/stanc/stanc.ml
@@ -49,7 +49,6 @@ let main () =
     | Compile settings -> settings in
   Debugging.lexer_logging := debug_lex;
   Debugging.grammar_logging := debug_parse;
-  Typechecker.model_name := Option.value ~default:"" name;
   Driver.Flags.set_backend_args_list
     (* remove executable itself from list before passing *)
     (Sys.get_argv () |> Array.to_list |> List.tl_exn);
@@ -65,7 +64,9 @@ let main () =
       , Option.first_some flags.filename_in_msg (Some "stdin") )
     else (model_file, `File model_file, flags.filename_in_msg) in
   match
-    Driver.Entry.stan2cpp model_file_name model_source flags
+    Driver.Entry.stan2cpp
+      (Option.value ~default:model_file_name name)
+      model_source flags
       (output_callback output_file printed_filename)
   with
   | Ok cpp_str ->

--- a/test/integration/cli-args/model-name/stanc.t
+++ b/test/integration/cli-args/model-name/stanc.t
@@ -1,0 +1,15 @@
+Test model name mangeling for non-alphanumeric characters
+  $ echo 'parameters {real y;}' >> añd.stan
+Filename
+  $ stanc añd.stan
+  $ grep "namespace a" añd.hpp
+  namespace ax195x177d_model_namespace {
+Name argument
+  $ stanc --name=añd añd.stan
+  $ grep "namespace a" añd.hpp
+  namespace ax195x177d_namespace {
+Name argument with dash
+  $ stanc --name="a-d" añd.stan
+  $ grep "namespace a" añd.hpp
+  namespace a_d_namespace {
+  $ rm añd.hpp añd.stan

--- a/test/stancjs/model_name.js
+++ b/test/stancjs/model_name.js
@@ -1,0 +1,29 @@
+var stanc = require('../../src/stancjs/stancjs.bc.js');
+var utils = require("./utils/utils.js");
+
+
+let basic_model = `
+parameters {
+	real y;
+}
+model {
+    y ~ std_normal();
+}
+`
+
+let basic_name1 = stanc.stanc("basic1", basic_model, []);
+utils.print_error(basic_name1);
+console.assert(basic_name1.result.includes("namespace basic1_namespace"), "Error: namespace not in C++ code");
+
+// test that the state doesn't persist
+
+let basic_name2 = stanc.stanc("basic2", basic_model, []);
+utils.print_error(basic_name2);
+console.assert(basic_name2.result.includes("namespace basic2_namespace"), "Error: namespace not in C++ code");
+
+
+// test that _model is added if the name is file-like
+
+let basic_name3 = stanc.stanc("basic.stan", basic_model, []);
+utils.print_error(basic_name3);
+console.assert(basic_name3.result.includes("namespace basic_model_namespace"), "Error: namespace not in C++ code");

--- a/test/stancjs/stancjs.expected
+++ b/test/stancjs/stancjs.expected
@@ -352,9 +352,7 @@ $ node info.js
 $ node math_sigs.js
 
 $ node model_name.js
-Assertion failed: Error: namespace not in C++ code
-Assertion failed: Error: namespace not in C++ code
-Assertion failed: Error: namespace not in C++ code
+
 $ node optimization.js
 Semantic error in 'string', line 3, column 8 to column 11:
    -------------------------------------------------

--- a/test/stancjs/stancjs.expected
+++ b/test/stancjs/stancjs.expected
@@ -351,6 +351,10 @@ $ node info.js
 }
 $ node math_sigs.js
 
+$ node model_name.js
+Assertion failed: Error: namespace not in C++ code
+Assertion failed: Error: namespace not in C++ code
+Assertion failed: Error: namespace not in C++ code
 $ node optimization.js
 Semantic error in 'string', line 3, column 8 to column 11:
    -------------------------------------------------

